cmd/provision/main.go

```diff
@@ -98,74 +114,130 @@ func main() {
-		httpPort:         mainflux.Env(envHTTPPort, defHTTPPort),
-		mfBSURL:          mainflux.Env(envMfBSURL, defMfBSURL),
-		mfBSWhiteListURL: mainflux.Env(envMfBSWhiteListURL, defMfWhiteListURL),
-		mfCertsURL:       mainflux.Env(envMfCertsURL, defMfCertsURL),
-		Config: provision.Config{
-			MFEmail:          mainflux.Env(envMfUser, defMfUser),
-			MFPass:           mainflux.Env(envMfPass, defMfPass),
-			PredefinedThings: strings.Split(mainflux.Env(envThingIDs, defThingIDs), ","),
-			X509Provision:    provisionX509,
-			BSProvision:      provisionBS,
-			AutoWhiteList:    autoWhiteList,
-			BSContent:        fmt.Sprintf(mainflux.Env(envBSContent, defBSContent), mainflux.Env(envMQTTURL, defMQTTURL)),
+	cfg := provision.Config{
+		Server: provision.ServiceConf{
+			LogLevel:       mainflux.Env(envLogLevel, defLogLevel),
+			CACerts:        mainflux.Env(envCACerts, defCACerts),
+			ServerCert:     mainflux.Env(envServerCert, defServerCert),
+			ServerKey:      mainflux.Env(envServerKey, defServerKey),
+			HTTPPort:       mainflux.Env(envHTTPPort, defHTTPPort),
+			MfBSURL:        mainflux.Env(envMfBSURL, defMfBSURL),
+			MfWhiteListURL: mainflux.Env(envMfBSWhiteListURL, defMfWhiteListURL),
+			MfCertsURL:     mainflux.Env(envMfCertsURL, defMfCertsURL),
+			MfUser:         mainflux.Env(envMfUser, defMfUser),
+			MfPass:         mainflux.Env(envMfPass, defMfPass),
+			MfApiKey:       mainflux.Env(envMfApiKey, defMfApiKey),
+			ThingsLocation: mainflux.Env(envThingsLocation, defThingsLocation),
+			UsersLocation:  mainflux.Env(envUsersLocation, defUsersLocation),
+			TLS:            tls,
+		},
+		Bootstrap: provision.Bootstrap{
+			X509Provision: provisionX509,
+			Provision:     provisionBS,
+			AutoWhiteList: autoWhiteList,
+			Content:       fmt.Sprintf(mainflux.Env(envBSContent, defBSContent), mainflux.Env(envMQTTURL, defMQTTURL)),
+		},
+
+		// This is default conf for provision if there is no config file
+		Channels: []provision.Channel{
+			{
+				Name:     "control-channel",
+				Metadata: map[string]interface{}{"type": "control"},
+			}, {
+				Name:     "data-channel",
+				Metadata: map[string]interface{}{"type": "data"},
+			},
+		},
+		Things: []provision.Thing{
+			{
+				Name:     "thing",
+				Metadata: map[string]interface{}{"externalID": "xxxxxx"},
+			},
 		},
 	}
-	thingSdkCfg := mfSDK.Config{
-		BaseURL:           mainflux.Env(envThingsLocation, defThingsLocation),
-		HTTPAdapterPrefix: "http",
-		MsgContentType:    "application/json",
-		TLSVerification:   tls,
-	}
-	thingsSDK := mfSDK.NewSDK(thingSdkCfg)
 
-	userSdkCfg := mfSDK.Config{
-		BaseURL:           mainflux.Env(envUsersLocation, defUsersLocation),
-		HTTPAdapterPrefix: "http",
-		MsgContentType:    "application/json",
-		TLSVerification:   tls,
+	cfg.File = mainflux.Env(envConfigFile, "")
+	if cfg.File == "" {
+		cfg.File = defConfigFile
+		if err != provision.Save(cfg, defConfigFile) {
+			return cfg, err
+		}
```

Comment:
Why not `cfg.File = mainflux.Env(envConfigFile, defConfigFile)` ?

Reply:
because if file is set in environment I don't want to overwrite it

Reply:
It's what this function does. Use ENVAR if defined or return the default value